### ISSUE TITLE
Feature Request: allow simulator log level, predicate, style, and type as simulator arguments

### ISSUE THREAD

**Issue by markOsipov** [lyndsey-ferguson/fastlane-plugin-test_center#286]
### Feature Request

Add an option, `simctl_args`, that allows extra arguments to be passed pre-launched simulators.

#### Motivation Behind Feature

Before tests start, we are using this fastlane script to set env variables in simulators and start streaming sim logs into file:
```
 `xcrun simctl spawn #{deviceId} launchctl setenv PROJECT_FOLDER_PATH #{project_folder_path}`
  command = "xcrun simctl spawn #{deviceId} log stream --level info --predicate 'subsystem == \"#{subsystem}\" && category == \"#{category}\"' --style json --type log"
  fork {
    puts "Start streaming system log for device #{deviceId}"
    exec(command, out: file)
  }
```

> If you love this fastlane plugin, consider sponsoring it or asking your company to sponsor it. I would really appreciate any
> gesture: https://github.com/sponsors/lyndsey-ferguson. 😍

#### Feature Description

Like `xcargs`, allow users to pre-launch the simulators and pass extra arguments to each simulator. 

#### Alternatives or Workarounds

None

**Comment by lyndsey-ferguson** [comment]
Hello @markOsipov thank you for this bug report. It is a very interesting scenario. In the change from [v3.11.5 to v3.11.6](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/compare/v3.11.5...v3.11.6) I changed the code that is responsible for collecting logs and that could have affected it.

Also, while I don't have a copy of your Fastfile, nor a full copy of the logs, I do see indications that you're running parallel tests.

There could be two issues:
1. The cloned simulators may not have the same environment variables (I don't know enough about this).
2. More likely, you have the `include_simulator_logs` option set to true, which makes multi_scan pre-launch the simulators, causing problems with what you have done. If so, you're already starting up the simulator and logging yourself, maybe don't use that option?

**Comment by markOsipov** [comment]
@lyndsey-ferguson Thank you for your response! 
There is our call of multi_scan. `include_simulator_logs` is set to false, so it shouldn't be the cause of the problem
```
multi_scan(
        workspace: workspace,
        scheme: scheme,
        clean: false,
        device: @config[:device_model],
        slack_channel: @config[:ui_tests_report_channel],
        only_testing: testList,
        skip_testing: skipTests,
        fail_build: false,
        skip_slack: false,
        include_simulator_logs: false,            
        parallel_testrun_count: @config[:ui_tests_parallel_testrun_count].to_i,
        try_count: @config[:ui_tests_try_count].to_i,
        xcargs: "ONLY_ACTIVE_ARCH=NO",
        result_bundle: true,
        suppress_xcode_output: true,
        derived_data_path: @config[:derived_data_path],
        test_without_building: @config[:test_without_building]
      )
```

I also tried to set `quit_simulators` or `pre_delete_cloned_simulators` (or both) to false and watch the results. Simulators seemed to stop failing after start. Tests didn't passed though, but this time that could be not because of simulators, so 
I will try to push this way further..

**Comment by lyndsey-ferguson** [comment]
This was changed in [fastlane 1.153.0](https://github.com/fastlane/fastlane/pull/16833/files#diff-b6a3ba6dff446741cb76248e99156be7R42) AND [fastlane-plugin-test_center 3.11.6](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/compare/v3.11.5...v3.11.6#diff-8e5624a2717b59358a99a62cd65433a3R49) to pre-boot simulators in order to properly collect logs. 

Yours were working because you were already pre-booting your simulator and probably the clone kept that state.

Note: each simulator will produce its own logs, so you don't need to pre-boot your simulator anymore, UNLESS, you really want to use the predicate to filter the logs that you get. If you do, then I would convert this into a feature request to allow you to pass a predicate to the logs that you want.

Then, I would put both changes into `scan` and `multi_scan`

**Comment by lyndsey-ferguson** [comment]
Thoughts on the above?

**Comment by markOsipov** [comment]
Yes, we want to keep only filtered logs of some custom events. So we'll glad to see this feature in future versions of test-center)

**Comment by lyndsey-ferguson** [comment]
Okay, thanks for confirming! I'll update this when I get a chance.

**Comment by lyndsey-ferguson** [comment]
So @markOsipov, I didn't remember how spawn works until I started looking into this issue. A couple of questions:

1. You're setting a environment variable of `PROJECT_FOLDER_PATH`. Is that needed?
1. You're asking for a stream of the log, is that needed, or can we use the collect method to get the logs that match? 
  If we can use collect, then I simply modify Scan's `[copy_archive](https://github.com/fastlane/fastlane/blob/8c9809ed26393626d72402d8c8fd98db918c80ca/fastlane_core/lib/fastlane_core/device_manager.rb#L349)` method to apply the filter that you've asked for.

**Comment by markOsipov** [comment]
1. PROJECT_FOLDER_PATH is needed because we want to save logs in project folder on CI
2. I think we need streaming, because we use this logs in our tests. We check created events in real time like this:
```
logService?.hasEvent(
                        DebugEvent(.plasticCardBlocked),
                        afterWaiting: 5,
                        inLastEventsAmount: 2
                    ) {
                        Matchers.checkThat(
                            "в метрику отправлено событие \(MetricsName.plasticCardBlocked)", $0)
                    }
```

**Comment by lyndsey-ferguson** [comment]
I'm not familiar with those events, how does it work?

**Comment by lyndsey-ferguson** [comment]
Would be acceptable if the logs were saved in the logs dir path that you give to scan/multi_scan?

**Comment by markOsipov** [comment]
> I'm not familiar with those events, how does it work?

Application sending metrics after certain users actions. 
We are checking that log contains event of sending metrics which we expect. 

So we need logfile to contain last events in real time

**Comment by markOsipov** [comment]
> Would be acceptable if the logs were saved in the logs dir path that you give to scan/multi_scan?

Yes, thats would be OK. Than you!

**Comment by lyndsey-ferguson** [comment]
> Application sending metrics after certain users actions.
> We are checking that log contains event of sending metrics which we expect.
>
> So we need logfile to contain last events in real time

Ok, so the application logs user actions. From what I understand, each log entry will have a timestamp. You write that you're checking that the log contains that event.

It is not clear to me why you need to respond to those events in real time. Do you have code that responds the moment that you see the event in the log to do something, and that _something_ has to be at the moment?

Or, can you not just parse the resultant log file to mark off when things happened?

What I'm getting at is: do you need to read this log as it is written, or can you collect the log, with the filters, and then process it after the tests have completed?

The second makes this easier (and faster) to add code to handle. The former is definitely possible, but adds more complexity to fastlane's `scan` (and `multi_scan`).

**Comment by lyndsey-ferguson** [comment]
More questions while I try to figure out how to provide a feature that solves your case, and the more generic case.

It seems that I could do a couple of things.

I could add a string option that lets you spawn various other processes in the simulator after it has spun up.

I could add a callback option, that calls your code with a device UDID so that you can run whatever command you want on the device (thus set the `PROJECT_FOLDER_PATH` for example).

What still remains unclear to me is if I should provide the option to other users to provide a callback that is forked, like what you're doing, in order to respond to messages in the log. Or, in other words, should I build in support to fork code like you're doing, or let you, in the callback, do the fork yourself.

For example.

```ruby
sim_callback = lambda do |simulator_device_udid|
  command = "xcrun simctl spawn #{simulator_device_udid} log stream --level info --predicate 'subsystem == \"#{subsystem}\" && category == \"#{category}\"' --style json --type log"
  fork {
    puts "Start streaming system log for device #{deviceId}"
    exec(command, out: file)
  }
end

multi_scan(
  project: File.absolute_path('../AtomicBoy/AtomicBoy.xcodeproj'),
  scheme: 'AtomicBoy',
  try_count: 3,
  batch_count: 4,
  fail_build: false,
  simulator_started_callback: sim_callback
)
```

Keep in mind that this would spawn a fork for every simulator started. So, if I provide 2 devices, and ask for 4 parallel test runs, I'll get a total of 4 forks for each parallel test run, 2 times for each device, that's already 8 forks, and then another 8 for each simulator callback -- 16 forks for your test run. Which may be fine.

However, is it worth offering users the streaming log case and wrap up the forking myself and allow devs to have a callback for the log stream? My thoughts initially are no, your case is special. It would be a fun project, but not sure I want to spend the time making it generalizable unless people actually want this.

@markOsipov it is still not clear to me why you need the streaming log. Do you have another process watching for particular strings in the log? If you _do_ have another process running, what does it do (send a Slack notification, create a bug, make the computer beep, stop the tests immediately, what?) ? 

Well, I've shared some of my thoughts around this, tell me what you think and give me more info on why you need the streaming log. 🙏

**Comment by markOsipov** [comment]
I believe forks will be overkill now, so if there are no other cases similar to our, there is no need to make fork support for now. 
I will discuss it with my team though, in case of someone has an idea how to use it, but quick answer is - Callback option seems to be ok.

We don't have any other processes that watching the log, the only one thing we use it for now - is metric events check. 
I'll try explain again, how it works for us: 

1. We starting test in which we simulating user doing some action in app (for example - user blocks his credit card)
2. This action leads to sending metrics request, and this event presents in simulator log
3. We need to check that metrics request was actually sent right now and with correct event description. 
4. We check 2-3 last events in simulator log and expecting to see there our desirable event :)
5. If check is successful, everything is ok and test will go on. If check is failed, test will fail right now and every other step is pointless. Thats why we cant wait until test is finished to check our log.

**Comment by lyndsey-ferguson** [comment]
Thanks for explaining the scenario!

**Comment by lyndsey-ferguson** [comment]
Released in [Sponsors-first](https://github.com/sponsors/lyndsey-ferguson) version of [v3.15.1](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/docs/feature_details/multi_scan.md). Available to the general public around Dec 18th, 2020!

> Any individual supporter at the Basic Supporter tier or higher will get early access, and all members of an organization sponsor at the Silver tier or higher will get that same access.

**Comment by lyndsey-ferguson** [comment]
Released in [v3.15.0](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/releases/tag/v3.15.0)
